### PR TITLE
722 team 01 team crud endpoints

### LINKED ISSUE #722: [TEAM-01] Team CRUD endpoints
## Overview
Let registered hackers form, join, leave, and delete teams.

## Goals
- Full team lifecycle
- Guard against the obvious foot-guns

## Tasks
- [ ] `createTeam`
- [ ] `joinTeam`
- [ ] `leaveTeam`
- [ ] `deleteTeam`
- [ ] Guards: can't delete another team by accident; can't join >1 team; can't join a team already at 4
- [ ] Unit tests; reference the schema

## References
- Auth pattern (`protectedProcedure` + error handling): https://github.com/hackwestern/hackwestern/blob/dev/src/server/api/routers/scavenger-hunt.ts#L277-L304
- Unit-test pattern for edge cases: https://github.com/hackwestern/hackwestern/blob/dev/src/server/api/routers/application.test.ts#L21-L60

## Notes
Build backend first.

### PR DESCRIPTION
**Pull request by LucasVand**
closes (Github Issue number)

# Quick Overview of Changes

- added team crud endpoints
- createTeam creates a new team if the user creating the team is not already in a team
- joinTeam adds a user to a team if the team is not full and the user is not already in a team
- leaveTeam leaves a users current team and deletes the team if it has 0 members after the user leaves
- deleteTeam deletes the users current team
- added tests to test most important errors and successes 

# Checklist

- [ ] If any frontend changes were made, include a screenshot/demo in the overview
- [ ] Checked all uses of changed component(s)/function(s)
- [x] If any changes were made to our backend services, at least one (happy path) unit test was added OR a `TODO` comment was added to create one.
- [x] Check that CI is passing.
- [ ] If no (re-)reviews after 1-2 days, ping the web leads on the discord to remind them to review the PR.

@lucianlavric @JasmineGu2

### CODE REVIEW

File: src/server/api/routers/teams.ts
Comment (lucianlavric) [inline src/server/api/routers/teams.ts:1]: Can we make createTeam add the users calling the operation to the team that they create

File: src/server/api/routers/teams.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
+import { TRPCError } from "@trpc/server";
+import { createTRPCRouter, protectedProcedure } from "../trpc";
+import z from "zod";
+import { db } from "~/server/db";
+import { teams, users } from "~/server/db/schema";
+import { count, eq } from "drizzle-orm";
+import { randomBytes } from "crypto";
+
+export const teamsRouter = createTRPCRouter({
+  createTeam: protectedProcedure
```
Comment (lucianlavric) [inline src/server/api/routers/teams.ts:10]: Can we make createTeam add the users calling the operation to the team that they create

File: src/server/api/routers/teams.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        message: "User not found",
+      });
+    }
+    if (!currentTeam.teamId) {
+      throw new TRPCError({
+        code: "BAD_REQUEST",
+        message: "Cannot delete team because user is not in a team",
+      });
+    }
+    await db
```
Comment (lucianlavric) [inline src/server/api/routers/teams.ts:155]: Can we actually delete the db for the delete team operation?

**Review by lucianlavric — CHANGES_REQUESTED** [review]

File: src/server/api/routers/teams.ts (hunk 2 above)
Comment (LucasVand) [reply src/server/api/routers/teams.ts:155]: lol

File: src/server/api/routers/teams.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+import z from "zod";
+import { db } from "~/server/db";
+import { teams, users } from "~/server/db/schema";
+import { count, eq } from "drizzle-orm";
+import { randomBytes } from "crypto";
+
+export const teamsRouter = createTRPCRouter({
+  createTeam: protectedProcedure
+    .input(z.object({ name: z.string() }))
+    .query(async ({ ctx, input }) => {
```
Comment (lucianlavric) [inline src/server/api/routers/teams.ts]: use .mutation instead of .query
.query is used to for get requests, .mutation is used to change data

**Review by lucianlavric — CHANGES_REQUESTED** [review]

**Comment by lucianlavric** [comment]
Current build is not working due to the any type access

./src/server/api/routers/teams.test.ts
34:18  Error: Unsafe member access .teamId on an `any` value.  @typescript-eslint/no-unsafe-member-access

**Review by lucianlavric — APPROVED** [review]
